File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/modloaders/FabricVersion.java
package net.kdt.pojavlaunch.modloaders;

import androidx.annotation.NonNull;

/**
 * Fabric Loaderのバージョン情報を表すクラス
 */
public class FabricVersion {
    public String version;
    public boolean stable;

    /**
     * Fabric Loaderのディスクリプタークラス
     */
    public static class LoaderDescriptor extends FabricVersion {
        public FabricVersion loader;

        /**
         * このLoaderDescriptorの文字列表現を返します。
         */
        @NonNull
        @Override
        public String toString() {
            return loader != null ? loader.toString() : "null";
        }
    }

    /**
     * このFabricVersionの文字列表現を返します。
     */
    @NonNull
    @Override
    public String toString() {
        return version;
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/feature/mod/modloader/FabricVersion.kt
package com.arata.yukarilauncher.feature.mod.modloader

/** Fabricローダーのバージョン情報を保持するクラス。 */
open class FabricVersion {
    /** バージョン文字列。 */
    var version: String? = null
    /** 安定版かどうか。 */
    var stable: Boolean = false

    /** Fabricローダーのローダー記述子を表すクラス。 */
    class LoaderDescriptor : FabricVersion() {
        /** ローダーのバージョン情報。 */
        var loader: FabricVersion? = null

        /** ローダーの文字列表現を返す。 @return ローダーバージョンの文字列、nullの場合は"null" */
        override fun toString(): String {
            return loader?.toString() ?: "null"
        }
    }

    /** バージョンの文字列表現を返す。 @return バージョン文字列、nullの場合は"null" */
    override fun toString(): String {
        return version ?: "null"
    }
}
